File: 剑指Offer/14.剪绳子.py
class Solution:
    def cuttingRope(self, n: int) -> int:
        if n < 2: return 0
        if n == 2: return 1
        if n == 3: return 2

        dpTable = [0 for i in range(n + 1)]
        dpTable[1] = 1
        dpTable[2] = 2
        dpTable[3] = 3

        for i in range(4, n + 1):
            maxProduct = 0
            for j in range(1, n // 2 + 1):
                maxProduct = max(maxProduct, dpTable[j] * dpTable[i - j])
            dpTable[i] = maxProduct

        return dpTable[n]

File: 剑指Offer/14.剪绳子.cpp
class Solution {
public:
    int cuttingRope(int n) {
        if (n < 2) return 0;
        if (n == 2) return 1;
        if (n == 3) return 2;

        vector<int> dpTable(n + 1, 0);
        dpTable[0] = 0;
        dpTable[1] = 1;
        dpTable[2] = 2;
        dpTable[3] = 3;

        int maxProduct;
        for (int i = 3; i <= n; ++i) {
            maxProduct = 0;
            for (int j = i; j <= n / 2; ++j) {
                maxProduct = max(maxProduct, dpTable[j] * dpTable[n - j]);
            }
            dpTable[i] = maxProduct;
        }
        return dpTable[n];
    }
};